I tested manually this change on my Fedora 29:

```
$ git clone ~/prog/python/master /tmp/cpython
$ cd /tmp/cpython
$ wget https://github.com/python/cpython/pull/10919.patch
$ git apply 10919.patch
$ ./configure --with-pydebug && make

$ wget https://bugs.python.org/file47211/user.py
$ sudo ./python user.py
['',
 '/tmp/cpython/Lib',
 '/tmp/cpython/build/lib.linux-x86_64-3.8-pydebug',
 '/usr/local/lib/python38.zip']
PARENT: child exited with status 0
```

The site module doesn't fail anymore. Test to ensure that site fails without the fix:

```
$ cd /tmp/cpython
$ git checkout .
$ sudo ./python user.py
Fatal Python error: initsite: Failed to import the site module
Traceback (most recent call last):
  File "/tmp/cpython/Lib/site.py", line 579, in <module>
    main()
  File "/tmp/cpython/Lib/site.py", line 565, in main
    known_paths = addusersitepackages(known_paths)
  File "/tmp/cpython/Lib/site.py", line 312, in addusersitepackages
    user_site = getusersitepackages()
  File "/tmp/cpython/Lib/site.py", line 297, in getusersitepackages
    userbase = getuserbase() # this will also set USER_BASE
  File "/tmp/cpython/Lib/site.py", line 286, in getuserbase
    USER_BASE = _getuserbase()
  File "/tmp/cpython/Lib/site.py", line 261, in _getuserbase
    return joinuser("~", ".local")
  File "/tmp/cpython/Lib/site.py", line 251, in joinuser
    return os.path.expanduser(os.path.join(*args))
  File "/tmp/cpython/Lib/posixpath.py", line 249, in expanduser
    userhome = pwd.getpwuid(os.getuid()).pw_dir
KeyError: 'getpwuid(): uid not found: 12345'
PARENT: child exited with status 256
```

Note: I had to re-clone CPython in /tmp since user 12345 is not allowed to access my $HOME directory.

--- a/Lib/test/test_posixpath.py
+++ b/Lib/test/test_posixpath.py
@@ -246,6 +246,13 @@ class PosixPathTest(unittest.TestCase):
 
     def test_expanduser_home_envvar(self):
         with support.EnvironmentVarGuard() as env:
+            env['HOME'] = '/home/victor'
+            self.assertEqual(posixpath.expanduser("~"), "/home/victor")
+
+            # expanduser() strips trailing slash
+            env['HOME'] = '/home/victor/'
+            self.assertEqual(posixpath.expanduser("~"), "/home/victor")
+
             for home in '/', '', '//', '///':
                 with self.subTest(home=home):
                     env['HOME'] = home
@@ -254,10 +261,7 @@ class PosixPathTest(unittest.TestCase):
                     self.assertEqual(posixpath.expanduser("~/foo"), "/foo")
 
     def test_expanduser_pwd(self):
-        try:
-            import pwd
-        except ImportError:
-            self.skipTest("need pwd module")
+        pwd = support.import_module('pwd')
 
         self.assertIsInstance(posixpath.expanduser("~/"), str)
         self.assertIsInstance(posixpath.expanduser(b"~/"), bytes)

--- a/Lib/test/test_site.py
+++ b/Lib/test/test_site.py
@@ -6,6 +6,7 @@ executing have not been removed.
 """
 import unittest
 import test.support
+from test import support
 from test.support import (captured_stderr, TESTFN, EnvironmentVarGuard,
                           change_cwd)
 import builtins
@@ -297,7 +298,8 @@ class HelperFunctionsTests(unittest.TestCase):
             self.assertTrue(user_site.startswith(user_base), user_site)
 
         with mock.patch('os.path.isdir', return_value=False) as mock_isdir, \
-             mock.patch.object(site, 'addsitedir') as mock_addsitedir:
+             mock.patch.object(site, 'addsitedir') as mock_addsitedir, \
+             support.swap_attr(site, 'ENABLE_USER_SITE', True):
 
             # addusersitepackages() must not add user_site to sys.path
             # if it is not an existing directory

--- a/Lib/test/test_site.py
+++ b/Lib/test/test_site.py
@@ -257,7 +257,7 @@ class HelperFunctionsTests(unittest.TestCase):
         # the call sets USER_BASE *and* USER_SITE
         self.assertEqual(site.USER_SITE, user_site)
         self.assertTrue(user_site.startswith(site.USER_BASE), user_site)
-        self.assertTrue(site.USER_BASE, site.getuserbase())
+        self.assertEqual(site.USER_BASE, site.getuserbase())
 
     def test_getsitepackages(self):
         site.PREFIXES = ['xoxo']

